man/man1/gdal_sieve.1
.TH "gdal_sieve" 1 "Wed Sep 16 2015" "GDAL" \" -*- nroff -*-
.ad l
.nh
.SH NAME
gdal_sieve \- gdal_sieve\&.py 
removes small raster polygons
.SH "SYNOPSIS"
.PP
.PP
.nf

gdal_sieve.py [-q] [-st threshold] [-4] [-8] [-o name=value]
           srcfile [-nomask] [-mask filename] [-of format] [dstfile]
.fi
.PP
.SH "DESCRIPTION"
.PP
The gdal_sieve\&.py script removes raster polygons smaller than a provided threshold size (in pixels) and replaces replaces them with the pixel value of the largest neighbour polygon\&. The result can be written back to the existing raster band, or copied into a new file\&.
.PP
Additional details on the algorithm are available in the GDALSieveFilter() docs\&.
.PP
.IP "\fB\fB-q\fP:\fP" 1c
The script runs in quiet mode\&. The progress monitor is suppressed and routine messages are not displayed\&.
.PP
.IP "\fB\fB-st\fP \fIthreshold\fP:\fP" 1c
Set the size threshold in pixels\&. Only raster polygons smaller than this size will be removed\&.
.PP
.IP "\fB\fB-o\fP \fIname=value\fP:\fP" 1c
Specify a special argument to the algorithm\&. Currently none are supported\&. 
.PP
.IP "\fB\fB-4\fP:\fP" 1c
Four connectedness should be used when determining polygons\&. That is diagonal pixels are not considered directly connected\&. This is the default\&. 
.PP
.IP "\fB\fB-8\fP:\fP" 1c
Eight connectedness should be used when determining polygons\&. That is diagonal pixels are considered directly connected\&. 
.PP
.IP "\fB\fIsrcfile\fP\fP" 1c
The source raster file used to identify target pixels\&. Only the first band is used\&.
.PP
.IP "\fB\fB-nomask\fP:\fP" 1c
Do not use the default validity mask for the input band (such as nodata, or alpha masks)\&. 
.PP
.IP "\fB\fB-mask\fP \fIfilename\fP:\fP" 1c
Use the first band of the specified file as a validity mask (zero is invalid, non-zero is valid)\&. 
.PP
.IP "\fB\fIdstfile\fP\fP" 1c
The new file to create with the filtered result\&. If not provided, the source band is updated in place\&.
.PP
.IP "\fB\fB-of\fP \fIformat\fP:\fP" 1c
Select the output format\&. The default is GeoTIFF (GTiff)\&. Use the short format name\&. 
.PP
.PP
.SH "AUTHORS"
.PP
Frank Warmerdam <warmerdam@pobox.com> 


NEWS
= GDAL/OGR 1.11.3 Release Notes = 

The 1.11.3 release is a bug fix release.

== Build (Unix) ==
 * Support custom buildflags in GRASS plugin (#5850)
 * GRASS 7.0.0 support (#5852)
 * fix detection of ECW SDK 5.1 on Mac (#5867, #5390)
 * Add support for Poppler >= 0.31.0 (#6118)

== Port ==
 * /vsitar/: remove useless validation test that prevents from opening valid .tar files (#5864)

== Algorithms ==
 * gdal_grid: fix crash in optimized mode with GCC 4.4 on 64bit (#5987)
 * Fix failure in GDALSuggestedWarpOut2() when computing output image size in case the top-left and bottom-right corners transform to the same point (#5980)
 * Rasterize: add check to avoid burning negative x values (#5641)

== GDAL core ==
 * GMLJP2 reader: add compatibility with OGC CRS URL as found in GMLJP2v2 (#5940)
 * GMLJP2: on reading, don't do axis inversation if there's an explicit axisName requesting easting, northing order (#5960)
 * GMLJP2: add missing rangeParameters element to validate against GMLJP2 schema (#5707)

== Utilities ==
 * gdalbuildvrt: fix potential crash when using -b switch (#6095)
 * gdalserver: fix compilation with recent GNU libc (such as in Arch Linux) (#6073)
 * gdalwarp: emit error message if file specified with -cutline cannot be opened (#5921)
 * gdalwarp: fix 1.11.2 regression when invoking several times gdalwarp with several input files on the same target file, and when the input files have a nodata setting: only the last input file was warped onto the target dataset (#5909)
 * ogr2ogr: fix crash with -clipdst when a reprojection fails before (#5973)
 * ogr_layer_algebra.py: for Update, Clip and Erase, only creates attribute of input layer by default (#5976)

== GDAL drivers ==

ECRG driver:
 * change subdataset definition to make sure that they only consist of frames of same scale (#6043)

ERS driver:
 * fix SetProjection() (#5840)

GeoRaster driver:
 * New SRID search (#5911)
 * Fix user-defined SRID issue (#5881)

GTiff driver:
 * fix GTiffDataset::IsBlockAvailable() wrong behaviour when compiling against internal libtiff, when a BigTIFF file has a TileByteCounts with LONG/4-byte counts and not LONG8 (#6001)
 * avoid generated corrupted  right-most and bottom-most tiles for 12-bit JPEG-compressed (#5971)
 * on reading attach PROJ4 extension node for Mercator_Auxiliary_Sphere / WebMercator (#5924)

HDF4 driver:
 * fix AnyTypeToDouble() to use proper type (int instead of long) to work with DFNT_INT32/DFNT_UINT32 on 64-bit Linux (#5965)

HF2 driver:
 * fix reading side of the driver to work on architectures where char is unsigned, like PPC or ARM (#6082)

NetCDF driver:
 * fix crash on opening a NOAA dataset (#5962)
 * fix computation of inverse flattening (#5858)

Northwoord driver:
 * Fix computation of intermediate color table values on non-Intel platforms (#6091)
 * NWT_GRD: don't advertize scale/offset as they are transparently applied in IReadBlock() (#5839).

RPFTOC driver:
 * add tweak for weird relative directory names in the A.TOC file (#5979)

TIL driver:
 * fix half pixel shift in geo registration (#5961)

VRT driver:
 * add more checks to CheckCompatibleForDatasetIO() to avoid issues with overview bands (#5954)
 * fix rounding of output window size on VRTSimpleSource (#5874)

== OGRSpatialReference ==
 * morphFromESRI(): map Mercator_Auxiliary_Sphere to EPSG:3857 (#5924)

== OGR drivers ==

EDIGEO driver:
 * backport conversion of atof() to CPLAtof() to avoid truncation of floating point numbers including coordinates on a non C locale (#6070)

GML driver:
 * accept choice between gml:polygonProperty and gml:multiPolygonProperty (#5977)
 * fix GML_ATTRIBUTES_TO_OGR_FIELDS=YES to work correctly with xlink:href too (#5970); and also backport a trunk fix for writing the correctly path to attributes in nested elements
 * VFR: add support for OriginalniHraniceOmpv (#5853)

GeoJSON driver:
 * avoid truncation of real numbers on reading (#5882)
 * TopoJSON: fix segfault when reading https://raw.githubusercontent.com/mbostock/topojson/master/test/topojson/properties-id-computed.json (#5919)
 * GeoJSON writer: make string comparison for authority name case insensitive so as to recognize lowercase 'epsg' (#4995)

GPKG driver:
 * on reading, recognize GEOMCOLLECTION to be compatible of GDAL 2.0. Still writing GEOMETRYCOLLECTION in 1.11 branch (#5937)
 * escape column names on reading/insert/update (#5879)

MSSQLSpatial driver:
 * Add support for creating features with default values (#5957)

OpenFileGDB driver:
 * fix 1.11.2 regression with FileGDB v9 where the presence of a non spatial table can cause other layers not to be listed (#5875)

PG/PGDump drivers:
 * fix truncation of fields to work with multi-byte UTF-8 characters (#5854)

Shape driver:
 * SHPReadOGRFeature() should free the passed psShape when not NULL and if the DBF record is deleted, so as to avoid later reads to faild (linked to #5974)

SQLite driver:
 * Fix issue with consecutive identical characters in layer name with ExecuteSQL() (#6107)

== SWIG Language Bindings ==

Perl bindings:
 * Fix creation of Field Definition object (#6099)

Python bindings:
 * Fix build with SWIG 3.0.6 (#6045)

= GDAL/OGR 1.11.2 Release Notes = 

The 1.11.2 release is a bug fix release.

== Build ==

  * configure: fix detection of OCI by changing linking order to please modern GCC (#5550)
  * configure: fix test to accept MariaDB 10.X as valid MySQL (#5722)
  * More compiler warnings addressed (#5414)
  * Windows build of PDF: fix compilation issue with Visual Studio 2012 (#5744)
  * Windows build: Add support for MrSID 9.1 SDK (#5814)
  * Windows build: when building netCDF, HDF4, HDF5 as plugins, call registration of 'sub-drivers' GMT, HDF4Image and HDF5Image (#5802)

== Port ==

 * CSV finder: Stop probing for csv/horiz_cs.csv. (#5698)
 * /vsicurl/: avoid reading after end-of-file and fix failure when reading more than 16MB in a single time (#5786)
 * CPLHexToBinary(): faster implementation (#5812)


swig/python/scripts/gdal_sieve.py
#!/usr/bin/env python
#******************************************************************************
#  $Id: gdal_sieve.py 18953 2010-02-28 12:00:54Z rouault $
# 
#  Project:  GDAL Python Interface
#  Purpose:  Application for applying sieve filter to raster data.
#  Author:   Frank Warmerdam, warmerdam@pobox.com
# 
#******************************************************************************
#  Copyright (c) 2008, Frank Warmerdam
# 
#  Permission is hereby granted, free of charge, to any person obtaining a
#  copy of this software and associated documentation files (the "Software"),
#  to deal in the Software without restriction, including without limitation
#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
#  and/or sell copies of the Software, and to permit persons to whom the
#  Software is furnished to do so, subject to the following conditions:
# 
#  The above copyright notice and this permission notice shall be included
#  in all copies or substantial portions of the Software.
# 
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
#  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#******************************************************************************

try:
    from osgeo import gdal, ogr
except ImportError:
    import gdal
    import ogr

import sys
import os.path

def Usage():
    print("""
gdal_sieve [-q] [-st threshold] [-4] [-8] [-o name=value]
           srcfile [-nomask] [-mask filename] [-of format] [dstfile]
""")
    sys.exit(1)
    
# =============================================================================
# 	Mainline
# =============================================================================

threshold = 2
connectedness = 4
options = []
quiet_flag = 0
src_filename = None

dst_filename = None
format = 'GTiff'

mask = 'default'

gdal.AllRegister()
argv = gdal.GeneralCmdLineProcessor( sys.argv )
if argv is None:
    sys.exit( 0 )

# Parse command line arguments.
i = 1
while i < len(argv):
    arg = argv[i]

    if arg == '-of':
        i = i + 1
        format = argv[i]

    elif arg == '-4':
        connectedness = 4
        
    elif arg == '-8':
        connectedness = 8
        
    elif arg == '-q' or arg == '-quiet':
        quiet_flag = 1
        
    elif arg == '-st':
        i = i + 1
        threshold = int(argv[i])
        
    elif arg == '-nomask':
        mask = 'none'
        
    elif arg == '-mask':
        i = i + 1
        mask = argv[i]
        
    elif arg == '-mask':
        i = i + 1
        mask = argv[i]
        
    elif arg[:2] == '-h':
        Usage()
        
    elif src_filename is None:
        src_filename = argv[i]

    elif dst_filename is None:
        dst_filename = argv[i]

    else:
        Usage()

    i = i + 1

if src_filename is None:
    Usage()
    
# =============================================================================
# 	Verify we have next gen bindings with the sievefilter method.
# =============================================================================
try:
    gdal.SieveFilter
except:
    print('')
    print('gdal.SieveFilter() not available.  You are likely using "old gen"')
    print('bindings or an older version of the next gen bindings.')
    print('')
    sys.exit(1)

# =============================================================================
#	Open source file
# =============================================================================

if dst_filename is None:
    src_ds = gdal.Open( src_filename, gdal.GA_Update )
else:
    src_ds = gdal.Open( src_filename, gdal.GA_ReadOnly )
    
if src_ds is None:
    print('Unable to open %s ' % src_filename)
    sys.exit(1)

srcband = src_ds.GetRasterBand(1)

if mask is 'default':
    maskband = srcband.GetMaskBand()
elif mask is 'none':
    maskband = None
else:
    mask_ds = gdal.Open( mask )
    maskband = mask_ds.GetRasterBand(1)

# =============================================================================
#       Create output file if one is specified.
# =============================================================================

if dst_filename is not None:

    drv = gdal.GetDriverByName(format)
    dst_ds = drv.Create( dst_filename,src_ds.RasterXSize, src_ds.RasterYSize,1,
                         srcband.DataType )
    wkt = src_ds.GetProjection()
    if wkt != '':
        dst_ds.SetProjection( wkt )
    dst_ds.SetGeoTransform( src_ds.GetGeoTransform() )
    
    dstband = dst_ds.GetRasterBand(1)
else:
    dstband = srcband

# =============================================================================
#	Invoke algorithm.
# =============================================================================

if quiet_flag:
    prog_func = None
else:
    prog_func = gdal.TermProgress
    
result = gdal.SieveFilter( srcband, maskband, dstband,
                           threshold, connectedness, 
                           callback = prog_func )
    









